### DISCUSSION TITLE
"count_distinct" `FunctionExpression`

### DISCUSSION THREAD

**Discussion by MarcoGorelli** [duckdb/duckdb#15485] [Ideas]
I'd like to be able to create an expression like
```
duckdb.FunctionExpression('count_distinct', duckdb.ColumnExpression('b'))
```

use-case: DuckDB support in [Narwhals](https://github.com/narwhals-dev/narwhals)

**Comment by soerenwolfers** [comment]
Not sure if this is what you mean (because it's more of a question of the pure SQL implementation -- the translation to Python is straightforward) but here goes:

```python
import duckdb
import pandas as pd
from duckdb import FunctionExpression, ColumnExpression

count_distinct = FunctionExpression('len', FunctionExpression('list_distinct', ColumnExpression('b')))
duckdb.df(pd.DataFrame({'b': [[1, 2, 2]]})).select(count_distinct)
```
```
┌───────────────────────┐
│ len(list_distinct(b)) │
│         int64         │
├───────────────────────┤
│                     2 │
└───────────────────────┘
```

Note that this ignores `NULL`s, e.g., it returns `0` on `[NULL]`.

**Reply by MarcoGorelli** [reply]
thanks @soerenwolfers ! i'm looking for a way to count nulls too, like Polars would do:
```python

In [10]: pl.Series([1, 2, None, None]).n_unique()
Out[10]: 3

In [11]: pl.Series([1, 2, None]).n_unique()
Out[11]: 3
```

Is there a way to write a duckdb expression to do that?

**Reply by soerenwolfers** [reply]
other than adding a separate expression that determines whether `NULL` is present, I don't think so.

Btw , I realized `list_unique` exists (but also ignores `NULL`s)

**Reply by MarcoGorelli** [reply]
ok, thanks

for now I'll raise `NotImplementedError`, hopefully they'll introduce something like `'count_distinct'` as an aggregation 🤞

**Reply by soerenwolfers** [reply]
Really? Are you concerned about the inefficiency of a two pass solution? I think that'd be drowned in the general inefficiency of lists in DuckDB

**Reply by MarcoGorelli** [reply]
I'm not concerned about a two-pass solution, but I think it's not yet available, right? As in, are you suggesting something like
```python
count_distinct = FunctionExpression('len', FunctionExpression('list_distinct', ColumnExpression('b'))) + FunctionExpression('bool_or', ColumnExpression('b').isnull()).cast('int8')
```
?

Because then, that would throw
```python
BinderException: Binder Error: Aggregates cannot be present in a Project relation!
```
and `WindowExpression` isn't available yet in the Python API for DuckDB

**Reply by soerenwolfers** [reply]
Don't use an aggregate function, use `list_contains` (although I can't check right now whether that handles NULLs the way you need). Also, do prefer `list_unique` over `len(list_distinct`

**Reply by soerenwolfers** [reply]
or simply swap `sum` for `list_sum` in your solution, which will definitely work
EDIT: Ignore this. I need to think and read more before writing comments

**Reply by MarcoGorelli** [reply]
I appreciate your help, I'm just not managing to write a solution, e.g.:
```python
In [30]: duckdb.df(pd.DataFrame({'b': [[1, 2, None]]})).select(FunctionExpression('list_unique', ColumnExpression('b')))
Out[30]:
┌────────────────┐
│ list_unique(b) │
│     uint64     │
├────────────────┤
│              2 │
└────────────────┘

In [31]: duckdb.df(pd.DataFrame({'b': [[1, 2, None]]})).select(FunctionExpression('list_unique', ColumnExpression('b').isnull()))
---------------------------------------------------------------------------
BinderException                           Traceback (most recent call last)
Cell In[31], line 1
----> 1 duckdb.df(pd.DataFrame({'b': [[1, 2, None]]})).select(FunctionExpression('list_unique', ColumnExpression('b').isnull()))

BinderException: Binder Error: No function matches the given name and argument types 'list_unique(BOOLEAN)'. You might need to add explicit type casts.
        Candidate functions:
        list_unique(ANY[]) -> UBIGINT
```

**Reply by soerenwolfers** [reply]
I shouldn't try to answer these on my phone, I was just adding confusion with some of my incorrect suggestions there. Here is something that works

```python
import duckdb
import pandas as pd
from duckdb import FunctionExpression, ColumnExpression, LambdaExpression

a = FunctionExpression('list_unique', ColumnExpression('b')) + (FunctionExpression('len', FunctionExpression('list_filter', ColumnExpression('b'), LambdaExpression('x', ColumnExpression('x').isnull()))) > 0).cast('int')
duckdb.df(pd.DataFrame({'b': [[1, 2, 2, None]]})).select(a)
```
```
┌───────────────────────────────────────────────────────────────────────────────────┐
│ (list_unique(b) + CAST((len(list_filter(b, (x -> (x IS NULL)))) > 0) AS INTEGER)) │
│                                      int128                                       │
├───────────────────────────────────────────────────────────────────────────────────┤
│                                                                                 3 │
└───────────────────────────────────────────────────────────────────────────────────┘
```

I feel with you, writing these is not fun ....

**Reply by MarcoGorelli** [reply]
that works wonderfully! thanks so much @soerenwolfers !

And now, to wait for the next DuckDB release so we can use `LambdaExpression` without needing the pre-release 🚀

**Comment by MarcoGorelli** [comment]
closing then, as we have a solution which isn't too bad and works well (https://github.com/narwhals-dev/narwhals/pull/1762/files)

thanks again Soeren for your help! lots of love for the relational api here, once `WindowExpression` drops we should be able to cover all the API surface we're aiming for 🚀
